C7/UIElements/RightClickMenu.cs

```diff
@@ -0,0 +1,174 @@
+	}
+
+	public Button AddItem(string text, Texture icon = null)
+	{
+		var button = new Button();
+		button.Text = text;
+		if (icon != null)
+			button.Icon = icon;
+		button.Align = Button.TextAlign.Left;
+		this.AddChild(button);
+		return button;
+	}
+
+	public void RemoveAll()
+	{
+		foreach (Node child in this.GetChildren())
+			child.QueueFree();
+	}
+
+	public override void _Input(InputEvent @event)
+	{
+		bool mouseOverMenu = new Rect2(Vector2.Zero, this.RectSize).HasPoint(this.GetLocalMousePosition()),
+		     escapeKeyWasPressed = (@event is InputEventKey keyEvent) && keyEvent.Pressed && keyEvent.Scancode == (int)Godot.KeyList.Escape,
+		     mouseClickedOutsideMenu = (@event is InputEventMouseButton mouseButtonEvent) && mouseButtonEvent.IsPressed() && ! mouseOverMenu;
+
+		if (escapeKeyWasPressed || mouseClickedOutsideMenu) {
+			this.AcceptEvent(); // Prevents other controls from receiving this event
+			CloseAndDelete();
+
+		// Eat all events other than mouse events while the cursor is over the menu. We want the menu to grab all input while it's open but we
+		// must make sure not to block mouse events from reaching its child buttons. (This had me confused for a while since the Godot docs
+		// say that events reach children before their parents, but the catch is that there are three phases of input processing. The "input"
+		// phase, this function, then "gui input", and finally "unhandled input". If a control eats an event during the "input" phase it won't
+		// proceed to the "gui input" phase where buttons actually respond to it.)
+		} else if (! ((@event is InputEventMouse) && mouseOverMenu))
+			this.AcceptEvent();
+	}
+}
+
+public class RightClickTileMenu : RightClickMenu
+{
+	public RightClickTileMenu(Game game, Tile tile) : base(game)
+	{
+		ResetItems(tile);
+	}
+
+	// assumeUnfortified is a unit that will be considered unfortified regardless of what its isFortified field holds. It's used to work around
+	// the fact that unfortification is done asynchronously through a message to the engine.
+	public void ResetItems(Tile tile, MapUnit assumeUnfortified = null)
+	{
+		RemoveAll();
+		foreach (MapUnit unit in tile.unitsOnTile) {
+			bool isFortified = (unit != assumeUnfortified) && unit.isFortified;
+			string action = (unit.owner == game.controller) ?
+				(isFortified ? "Wake" : "Activate") :
+				"Contact";
+			AddItem($"{action} {unit.Describe()}").Connect("pressed", this, "SelectUnit", new Godot.Collections.Array() {unit.guid});
+		}
+	}
+
+	public void SelectUnit(string guid)
+	{
+		using (var gameDataAccess = new UIGameDataAccess()) {
+			MapUnit toSelect = gameDataAccess.gameData.mapUnits.Find(u => u.guid == guid);
+			if (toSelect != null && toSelect.owner == game.controller) {
+				game.setSelectedUnit(toSelect);
+				new MsgSetFortification(toSelect.guid, false).send();
+				ResetItems(toSelect.location, toSelect);
+			}
+		}
+		if (! Input.IsKeyPressed((int)Godot.KeyList.Shift))
+			CloseAndDelete();
```

Comment:
Whoa, I never knew that you could keep the menu open and de-fortify multiple units by holding down the Shift key!  I always thought it was either all of them or one at a time.  Probably wouldn't have realized it was possible in C7 either, if it weren't for reading the code...